src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp

```diff
@@ -1095,8 +1095,15 @@ class MacroAssembler: public Assembler {
     return CodeCache::max_distance_to_non_nmethod() > branch_range;
   }
 
-  // Jumps that can reach anywhere in the code cache.
-  // Trashes tmp.
+  // Far_call and far_jump generate a call of/jump to the provided address.
+  // The address must be inside the code cache.
+  // Supported entry.rspec():
+  // - relocInfo::external_word_type
+  // - relocInfo::runtime_call_type
+  // - relocInfo::none
+  // If the distance to the address can exceed the branch range (128M for the release build)
+  // for direct calls/jumps, an indirect call/jump is generated.
```

Comment:
What is `indirect` in this case? Is it the use of `adrp` or it is `trampoline` (which is also kind of `indirect`)?
What is used in `debug` build (since you mentioned `release` build)?
Would be nice if you also add similar extended description for `trampoline_call()`.

Reply:
> Would be nice if you also add similar extended description for trampoline_call().

Yes, I'll do: https://bugs.openjdk.java.net/browse/JDK-8286058

> What is indirect in this case? Is it the use of adrp or it is trampoline (which is also kind of indirect)?

It is a call/branch at an address in a register. Yes, a trampoline is an indirect far call. I'll update the comment.
Trampolines use `ldr` instead of `adrp` which allows jumping anywhere in the memory. `adrp` in limited to 4Gb. As trampolines are always used for calling code inside `CodeCache` and `CodeCache` is limited to 2Gb, using `ldr` might be expensive in such cases. Created https://bugs.openjdk.java.net/browse/JDK-8286140.

> What is used in debug build (since you mentioned release build)?

2M. 128M is ISA limit.